### PR TITLE
[WIP] Counter-based gather

### PR DESCRIPTION
**Pull request by luizirber**
This is a Python-level implementation of what [greyhound](https://github.com/dib-lab/sourmash/issues/1238) does for `gather`. I exposed as another method in the `Index` abc, with a blanket (and not very optimized...) implementation that depends on the `Index.signatures()` method.

## TODO

- [ ] SBT supports a faster impl (using `search`), but needs some way of indexing signatures by identifier (I used the index in the signature list as index in greyhound/blanket impl, but that forces loading all the signatures from disk in the SBT...)
- [ ] For testing I replaced the call to `obj.gather` with `obj.counter_gather` in `src/sourmash/search.py`, but that's sub-optimal (because the `gather` in `search.py` redoes a lot of the work that `counter_gather` is doing). Need to move more functionality around to avoid redoing the work

## Checklist

- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

**Comment by ctb** [comment]
Digging in a bit more, what about splitting the `counter_gather` function in two - one that does the counter building (the prefetch), and one that uses it to do the gather?

**Comment by ctb** [comment]
Looking at this code, I can't help but think there are strong similarities with the `_find_signatures` method in `LCA_Database`.

**Comment by ctb** [comment]
@luizirber, see `CounterGatherIndex` in https://github.com/dib-lab/sourmash/pull/1371/ for a mashup of the "counter gather" approach with a separate `prefetch` function. There are still some problems with the tests that I haven't tracked down yet, but it mostly works. Curious what you think of the approach.

File: src/sourmash/index.py [hunk 1]
```diff
@@ -117,6 +117,69 @@ def gather(self, query, *args, **kwargs):
...
+            match = signatures[dataset_id]
+            del counter[dataset_id]
+            cont = query.minhash.contained_by(match.minhash, True)
+            if cont and cont >= threshold:
+                results.append((cont, match, getattr(self, "filename", None)))
+
+            # Prepare counter for finding the next match by decrementing
+            # all hashes found in the current match in other datasets
+            for (dataset_id, _) in most_common:
+                counter[dataset_id] -= signatures[dataset_id].minhash.count_common(match.minhash, True)
```
Comment (ctb) [inline src/sourmash/index.py:175]: Whups, this subtraction needs to be done for the overlap with _intersection_ of the match and the query, not the overlap with the query (which may be far larger).

Comment (ctb) [reply src/sourmash/index.py:175]: (see the copypasta code in https://github.com/dib-lab/sourmash/pull/1371, `src/sourmash/index.py::CounterGatherIndex.gather(...)`, which I fixed to pass gather tests)

**Comment by ctb** [comment]
should this be closed, now that #1370 has been merged?

**Comment by luizirber** [comment]
Yup.

**Pull request closed without merging**